core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java

```diff
@@ -1191,6 +1202,12 @@ private static QueryStats pruneQueryStats(QueryStats queryStats)
         @GuardedBy("this")
         private boolean noMoreExchangeLocations;
 
+        private final Object lock = new Object();
```

Comment:
why do we need artificial lock. Use `this` or comment why different lock used.

Reply:
This is to avoid synchronizing on the same lock as the other methods in this class to avoid unnecessary contention.

Reply:
I think having multiple locks increses complexity and should be introduced only if simpler locking mechanism show problematic. https://www.azquotes.com/picture-quotes/quote-premature-optimization-is-the-root-of-all-evil-in-programming-tony-hoare-67-37-76.jpg ;)

Reply:
And if you really insist to keep the another lock it should be named more specifically. Like `dynamicFiltersStatsSupplierLock` which we have already.

Reply:
Removed the "lock" object